### PR TITLE
KON-428: Process clients

### PR DESCRIPTION
**Pull request by rimi-itk**
https://jira.itkdev.dk/browse/KON-428

Adds person and company process clients.

### CODE REVIEW

File: CVR/ServiceplatformenCvrServiceResult.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        $address = $this->getStreetName();
+
+        $address .= null !== $this->getHouseNumber()
+            ? ' '.$this->getHouseNumber()
+            : ''
+        ;
+
+        return $address;
+    }
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply CVR/ServiceplatformenCvrServiceResult.php:88]: Good point! I've lifted the result class from https://github.com/itk-dev/os2forms_cvr_lookup/blob/develop/src/CVR/CvrServiceResult.php which may also needs to be extended …

File: Export/Reports/RevenueExport.php [hunk 2]
```diff
@@ -195,7 +195,9 @@ private function newEntry()
     private function getProcesses(): Traversable
     {
         $queryBuilder = $this->entityManager->getRepository(Process::class)->createQueryBuilder('p')
-            ->andWhere('p.originallyCompletedAt IS NOT NULL');
+            ->andWhere('p.originallyCompletedAt IS NOT NULL')
+            ->join('p.processClient', 'client')
+            ->andWhere('client.type = \'person\'');
```
Comment (lats1) [inline Export/Reports/RevenueExport.php:200]: Should the `ProcessClientPerson::TYPE` constant be used here?

**Review by lats1 — CHANGES_REQUESTED** [review]

File: CPR/ServiceplatformenServiceFactory.php [hunk 3]
```diff
@@ -75,23 +82,24 @@ public static function createPersonBaseDataExtendedService(string $azureTenantId
...
 
-        if (!realpath($personBaseDataExtendedServiceContract)) {
-            throw new CprException('The path to the service contract is invalid.');
+        if (!realpath($serviceContractFilename)) {
+            throw new CprException(sprintf('The path (%s) to the service contract is invalid.', $serviceContractFilename));
         }
 
         try {
-            $soapClient = new \SoapClient($personBaseDataExtendedServiceContract, $options);
+            $options['trace'] = true;
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply CPR/ServiceplatformenServiceFactory.php:93]: This looks like a leftover from a debugging session. I'll remove it.

File: Controller/ProcessController.php [hunk 4]
```diff
@@ -170,27 +185,62 @@ public function index(Request $request, ProcessRepository $processRepository, Fi
...
+            try {
+                $client = $clientManager->createClient($clientType, $clientData);
+                // Populate the client to catch errors in (or missing) client data.
+                $clientManager->populateClient($client);
+                $process = $processManager->newProcess($process, $client);
+                $processCreated = true;
+            } catch (\Exception $exception) {
+                $identifierName = ProcessClientPerson::TYPE === $clientType ? 'cpr' : 'cvr';
+                if ($form->has($clientType) && $form->get($clientType)->has($identifierName)) {
+                    $form->get($clientType)->get($identifierName)->addError(new FormError($exception->getMessage()));
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply Controller/ProcessController.php:235]: Good and valid point, but `ProcessClientManager::populateClient` will throw either a `CprException` or `CvrException` which have no common superclass apart from `Exception`. We could create a new superclass, but I'm not sure it is worth the effort.

File: Export/Reports/RevenueExport.php (hunk 2 above)
Comment (rimi-itk) [reply Export/Reports/RevenueExport.php:200]: Yes, of course.

File: Controller/ProcessController.php [hunk 5]
```diff
@@ -170,27 +185,62 @@ public function index(Request $request, ProcessRepository $processRepository, Fi
...
      * @throws \Doctrine\ORM\NonUniqueResultException
+     * @throws \Kontrolgruppen\CoreBundle\CPR\CprException
      */
-    public function new(Request $request, ProcessManager $processManager): Response
+    public function new(Request $request, ProcessManager $processManager, ProcessClientManager $clientManager): Response
     {
+        // Force user to select process client type before anything else.
+        try {
+            $client = $clientManager->createClient($request->get('clientType') ?? '');
+        } catch (\Exception $exception) {
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply Controller/ProcessController.php:203]: We, currently, don't have a common exception superclass to catch.

File: Controller/ProcessController.php [hunk 6]
```diff
@@ -170,27 +185,62 @@ public function index(Request $request, ProcessRepository $processRepository, Fi
...
+            $this->addFlash('danger', $exception->getMessage());
+
+            return $this->render('process/select-client-type.html.twig');
+        }
+
         $process = new Process();
+        $process->setProcessClient($client);
 
         $this->denyAccessUnlessGranted('edit', $process);
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply Controller/ProcessController.php:212]: It probably should. I'll fix it.

File: Controller/ProcessController.php [hunk 7]
```diff
@@ -213,37 +263,73 @@ public function new(Request $request, ProcessManager $processManager): Response
...
         }
 
-        $processes = $processRepository->findBy(
-            ['clientCPR' => $request->request->get('cpr')]
-        );
+        $processes = [];
+        $clients = $clientRepository->findBy(['identifier' => $cpr]);
+        foreach ($clients as $client) {
+            $processes[] = $client->getProcess();
+        }
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply Controller/ProcessController.php:287]: Maybe. Should we move it to `ProcessRepository`?

[the remark this answers is not in the document]
Comment (rimi-itk) [reply Controller/ProcessController.php]: Fixed in https://github.com/itk-dev/kontrolgruppen-core-bundle/pull/244/commits/3a1fe9d1d100be7ce2747e878d04a936c61a30d7.

File: Controller/ProcessController.php [hunk 8]
```diff
@@ -213,37 +263,73 @@ public function new(Request $request, ProcessManager $processManager): Response
...
+        $cvr = $request->get('cvr');
+        if (!$cvr) {
+            throw new NotFoundHttpException('No CVR found!');
+        }
+
+        $processes = [];
+        $clients = $clientRepository->findBy(['identifier' => $cvr]);
+        foreach ($clients as $client) {
+            $processes[] = $client->getProcess();
+        }
```
[the remark this answers is not in the document]
Comment (rimi-itk) [reply Controller/ProcessController.php]: Fixed in https://github.com/itk-dev/kontrolgruppen-core-bundle/pull/244/commits/3a1fe9d1d100be7ce2747e878d04a936c61a30d7.

**Review by lats1 — APPROVED** [review]

**Pull request merged**